traversal/planner/GraphPlanner.java

```diff
@@ -231,37 +233,43 @@ Optimiser optimiser() {
     }
 
     void mayOptimise(GraphManager graphMgr, boolean singleUse) {
-        if (procedure == null) {
-            try {
-                firstOptimisingLock.writeLock().lock();
-                if (procedure == null) optimise(graphMgr, singleUse);
-                assert procedure != null;
-            } finally {
-                firstOptimisingLock.writeLock().unlock();
-            }
-        } else if (isOptimising.compareAndSet(false, true)) {
-            try {
-                optimise(graphMgr, singleUse);
-            } finally {
-                isOptimising.set(false);
-            }
+        long timeLimitMillis = singleUse ? HIGHER_TIME_LIMIT_MILLIS : DEFAULT_TIME_LIMIT_MILLIS;
+        if (backgroundOptimisation == null) beginFirstOptimise(graphMgr, timeLimitMillis);
+        else if (isOptimising.compareAndSet(false, true)) beginReOptimise(graphMgr, timeLimitMillis);
+
+        try {
+            backgroundOptimisation.get(timeLimitMillis, MILLISECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException ignored) {
+            LOG.trace("Query plan did not finish updating in " + timeLimitMillis + " ms.");
+        }
+    }
+
+    private synchronized void beginFirstOptimise(GraphManager graphMgr, long timeLimitMillis) {
+        if (backgroundOptimisation == null) {
+            isOptimising.set(true);
+            updateTraversalCosts(graphMgr);
+            updateOptimiser();
+            createProcedure();
```

Comment:
create an initial `procedure` based on the greedy initialisation that is the fallback in case the first call to `optimise()` takes much longer than expected